File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/NTAG215.java
package com.hiddenramblings.tagmo.nfctech;

import android.nfc.Tag;
import android.nfc.tech.MifareUltralight;
import android.nfc.tech.NfcA;
import android.nfc.tech.TagTechnology;

import com.hiddenramblings.tagmo.eightbit.io.Debug;

import java.io.IOException;

public class NTAG215 implements TagTechnology {

    private static final int NXP_MANUFACTURER_ID = 0x04;
    private static final int MAX_PAGE_COUNT = 256;

    private final MifareUltralight m_mifare;
    private final NfcA m_nfcA;
    private final int maxTransceiveLength;

    public NTAG215(NfcA nfcA) {
        m_nfcA = nfcA;
        m_mifare = null;
        maxTransceiveLength = (m_nfcA.getMaxTransceiveLength() / 4) + 1;
    }

    public NTAG215(MifareUltralight mifare) {
        m_nfcA = null;
        m_mifare = mifare;
        maxTransceiveLength = (m_mifare.getMaxTransceiveLength() / 4) + 1;
    }

    public static NTAG215 get(Tag tag) {
        NfcA nfcA = NfcA.get(tag);
        if (null != nfcA) {
            if (nfcA.getSak() == 0x00 && tag.getId()[0] == NXP_MANUFACTURER_ID)
                return new NTAG215(nfcA);
        }
        MifareUltralight mifare = MifareUltralight.get(tag);
        if (null != mifare)
            return new NTAG215(mifare);

        return null;
    }

    public static NTAG215 getBlind(Tag tag) {
        NfcA nfcA = NfcA.get(tag);
        if (null != nfcA) {
            return new NTAG215(nfcA);
        }
        return null;
    }

    @SuppressWarnings("unused")
    public int getTimeout() {
        if (null != m_mifare)
            return m_mifare.getTimeout();
        if (null != m_nfcA) {
            return m_nfcA.getTimeout();
        }
        return 0;
    }

    @SuppressWarnings("unused")
    public void setTimeout(int timeout) {
        if (null != m_mifare)
            m_mifare.setTimeout(timeout);
        if (null != m_nfcA) {
            m_nfcA.setTimeout(timeout);
        }
    }

    public byte[] transceive(byte[] data) {
        try {
            if (null != m_mifare) {
                return m_mifare.transceive(data);
            } else if (null != m_nfcA) {
                return m_nfcA.transceive(data);
            }
        } catch (IOException e) {
            Debug.Warn(e);
        }
        return null;
    }

    public byte[] readPages(int pageOffset) throws IOException {
        if (null != m_mifare)
            return m_mifare.readPages(pageOffset);
        else if (null != m_nfcA) {
            validatePageIndex(pageOffset);
            //checkConnected();

            byte[] cmd = {
                    NfcByte.CMD_READ,
                    (byte) pageOffset
            };
            return m_nfcA.transceive(cmd);
        }
        return null;
    }

    public void writePage(int pageOffset, byte[] data) throws IOException {
        if (null != m_mifare) {
            m_mifare.writePage(pageOffset, data);
        } else if (null != m_nfcA) {
            validatePageIndex(pageOffset);
            //m_nfcA.checkConnected();

            byte[] cmd = new byte[data.length + 2];
            cmd[0] = (byte) NfcByte.CMD_WRITE;
            cmd[1] = (byte) pageOffset;
            System.arraycopy(data, 0, cmd, 2, data.length);

            m_nfcA.transceive(cmd);
        }
    }

    private static void validatePageIndex(int pageIndex) {
        // Do not be too strict on upper bounds checking, since some cards
        // may have more addressable memory than they report.
        // Note that issuing a command to an out-of-bounds block is safe - the
        // tag will wrap the read to an addressable area. This validation is a
        // helper to guard against obvious programming mistakes.
        if (pageIndex < 0 || pageIndex >= MAX_PAGE_COUNT) {
            throw new IndexOutOfBoundsException("page out of bounds: " + pageIndex);
        }
    }

    public void connect() throws IOException {
        if (null != m_mifare) {
            m_mifare.connect();
        } else if (null != m_nfcA) {
            m_nfcA.connect();
        }
    }

    public void close() throws IOException {
        if (null != m_mifare) {
            m_mifare.close();
        } else if (null != m_nfcA) {
            m_nfcA.close();
        }
    }

    public Tag getTag() {
        if (null != m_mifare) {
            return m_mifare.getTag();
        } else if (null != m_nfcA) {
            return m_nfcA.getTag();
        }
        return null;
    }

    /*
     * byte 1: currently active slot
     * byte 2: number of active banks
     * byte 3: button pressed?
     * byte 4: FW version?
     * see: http://wiki.yobi.be/wiki/N2_Elite#0x55:_N2_GET_INFO
     */
    public byte[] getVersion(boolean isGeneric) {
        byte[] command = isGeneric
                ? new byte[]{ NfcByte.CMD_GET_VERSION }
                : new byte[]{ NfcByte.N2_GET_VERSION };
        return this.transceive(command);
    }

    @SuppressWarnings("unused")
    public byte[] getBankCount() {
        byte[] req = new byte [1];
        byte[] resp;

        req[0] = NfcByte.N2_BANK_COUNT;

        return this.transceive(req);
    }

    public byte[] readSignature(boolean isGeneric) {
        byte[] command = isGeneric
                ? new byte[] { NfcByte.CMD_READ_SIG, (byte) 0x00 }
                : new byte[]{ NfcByte.N2_READ_SIG };
        return this.transceive(command);
    }

    public void setBankCount(int count) {
        this.transceive(new byte[]{
                (byte) NfcByte.N2_SET_BANKCOUNT,
                (byte) (count & 0xFF)
        });
    }

    public void activateBank(int bank) {
        this.transceive(new byte[]{
                (byte) NfcByte.N2_ACTIVATE_BANK,
                (byte) (bank & 0xFF)
        });
    }

    public void initFirmware() {
        this.transceive(new byte[]{
                (byte) 0xFFF4, (byte) 0x49,
                (byte) 0xFF9B, (byte) 0xFF99,
                (byte) 0xFFC3, (byte) 0xFFDA,
                (byte) 0x57, (byte) 0x71,
                (byte) 0x0A, (byte) 0x64,
                (byte) 0x4A, (byte) 0xFF9E,
                (byte) 0xFFF8, (byte) NfcByte.CMD_WRITE,
                (byte) NfcByte.CMD_READ, (byte) 0xFFD9
        });
    }

    private interface IFastRead {
        byte[] doFastRead(int i, int i2, int i3);
    }

    private interface IFastWrite {
        boolean doFastWrite(int i, int i2, byte[] bArr);
    }

    public byte[] fastRead(int startAddr, int endAddr) {
        return internalFastRead((startAddr1, endAddr1, bank) -> this.transceive(new byte[]{
                NfcByte.CMD_FAST_READ,
                (byte) (startAddr1 & 0xFF),
                (byte) (endAddr1 & 0xFF)
        }), startAddr, endAddr, 0);
    }

    public byte[] amiiboFastRead(int startAddr, int endAddr, int bank) {
        return internalFastRead((startAddr1, endAddr1, bank1) -> this.transceive(new byte[]{
                NfcByte.N2_FAST_READ,
                (byte) (startAddr1 & 0xFF),
                (byte) (endAddr1 & 0xFF),
                (byte) (bank1 & 0xFF)
        }), startAddr, endAddr, bank);
    }

    private byte[] internalFastRead(IFastRead method, int startAddr, int endAddr, int bank) {
        if (endAddr < startAddr) {
            return null;
        }
        byte[] resp = new byte[(((endAddr - startAddr) + 1) * 4)];
        int maxReadLength = (maxTransceiveLength / 4) - 1;
        if (maxReadLength < 1) {
            return null;
        }
        int snippetByteSize = maxReadLength * 4;
        int startSnippet = startAddr;
        int i = 0;
        while (startSnippet <= endAddr) {
            int endSnippet = (startSnippet + maxReadLength) - 1;
            if (endSnippet > endAddr) {
                endSnippet = endAddr;
            }
            byte[] respSnippet = method.doFastRead(startSnippet, endSnippet, bank);
            if (null == respSnippet) {
                return null;
            }
            if (respSnippet.length != ((endSnippet - startSnippet) + 1) * 4) {
                return null;
            }
            if (respSnippet.length == resp.length) {
                return respSnippet;
            }
            System.arraycopy(respSnippet, 0, resp, i * snippetByteSize, respSnippet.length);
            startSnippet += maxReadLength;
            i++;
        }
        return resp;
    }

    private boolean internalWrite(IFastWrite method, int addr, int bank, byte[] data) {
        byte[] query = new byte[4];
        int pages = data.length / 4;
        for (int i = 0; i < pages; i++) {
            System.arraycopy(data, i * 4, query, 0, 4);
            if (!method.doFastWrite(addr + i, bank, query)) {
                return false;
            }
        }
        return true;
    }

    public boolean amiiboWrite(int addr, int bank, byte[] data) {
        if (null != data && data.length % 4 == 0) {
            return internalWrite((startAddr, bank1, data1) -> {
                byte[] req = new byte[7];
                req[0] = (byte) NfcByte.N2_WRITE;
                req[1] = (byte) (startAddr & 0xFF);
                req[2] = (byte) (bank1 & 0xFF);
                try {
                    System.arraycopy(data1, 0, req, 3, 4);
                    this.transceive(req);
                    return true;
                } catch (Exception e) {
                    return false;
                }
            }, addr, bank, data);
        }
        return false;
    }

    private boolean internalFastWrite(IFastWrite method, int startAddr, int bank, byte[] data) {
        int snippetByteSize = 16;
        int endAddr = startAddr + (data.length / 4);
        int startSnippet = startAddr;
        int i = 0;
        while (startSnippet <= endAddr) {
            if (startSnippet + 4 >= endAddr) {
                snippetByteSize = data.length % snippetByteSize;
            }
            if (snippetByteSize == 0) {
                return true;
            }
            byte[] query = new byte[snippetByteSize];
            System.arraycopy(data, i, query, 0, snippetByteSize);
            if (!method.doFastWrite(startSnippet, bank, query)) {
                return false;
            }
            startSnippet += 4;
            i += snippetByteSize;
        }
        return true;
    }

    public boolean amiiboFastWrite(int addr, int bank, byte[] data) {
        if (null == data) {
            return false;
        }
        return internalFastWrite((startAddr, bank1, data1) -> {
            byte[] req = new byte[(data1.length + 4)];
            req[0] = (byte) NfcByte.N2_FAST_WRITE;
            req[1] = (byte) (startAddr & 0xFF);
            req[2] = (byte) (bank1 & 0xFF);
            req[3] = (byte) (data1.length & 0xFF);
            try {
                System.arraycopy(data1, 0, req, 4, data1.length);
                this.transceive(req);
                return true;
            } catch (Exception e) {
                return false;
            }
        }, addr, bank, data);
    }

    public void amiiboLock() {
        this.transceive(new byte[]{
                NfcByte.N2_LOCK
        });
    }

    public byte[] amiiboPrepareUnlock() {
        return this.transceive(new byte[]{
                NfcByte.N2_UNLOCK_1
        });
    }

    public void amiiboUnlock() {
        this.transceive(new byte[]{
                NfcByte.N2_UNLOCK_2
        });
    }

    @Override
    public boolean isConnected() {
        return m_nfcA.isConnected();
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/NTAG215.kt
/*
 * ====================================================================
 * N2 Elite Copyright (C) 2019 MasterCheaterz
 * Copyright (c) 2021 AbandonedCart @ TagMo
 * ====================================================================
 */
package com.hiddenramblings.tagmo.nfctech

import android.nfc.Tag
import android.nfc.tech.MifareUltralight
import android.nfc.tech.NfcA
import android.nfc.tech.TagTechnology
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.eightbit.io.Debug
import java.io.IOException

class NTAG215 : TagTechnology {

    private val tagMifare: MifareUltralight?
    private val tagNfcA: NfcA?
    private var maxTransceiveLength: Int = 0

    constructor(nfcA: NfcA?) {
        tagNfcA = nfcA?.also {
            maxTransceiveLength = it.maxTransceiveLength / 4 + 1
        }
        tagMifare = null
    }

    constructor(mifare: MifareUltralight?) {
        tagNfcA = null
        tagMifare = mifare?.also {
            maxTransceiveLength = it.maxTransceiveLength / 4 + 1
        }
    }

    @Suppress("unused")
    var timeout: Int
        get() = tagMifare?.timeout ?: tagNfcA?.timeout ?: 0
        set(timeout) {
            tagMifare?.timeout = timeout
            tagNfcA?.timeout = timeout
        }

    @Throws(IOException::class)
    override fun connect() {
        tagMifare?.connect() ?: tagNfcA?.connect()
    }

    override fun isConnected(): Boolean {
        return tagMifare?.isConnected ?: tagNfcA?.isConnected ?: false
    }

    @Throws(IOException::class)
    override fun close() {
        tagMifare?.close() ?: tagNfcA?.close()
    }

    override fun getTag(): Tag? {
        return tagMifare?.tag ?: tagNfcA?.tag
    }

    fun transceive(data: ByteArray?): ByteArray? {
        return try {
            tagMifare?.transceive(data) ?: tagNfcA?.transceive(data)
        } catch (e: IOException) {
            Debug.warn(e)
            null
        }
    }

    fun getVersion(isGeneric: Boolean): ByteArray? {
        val command = if (isGeneric)
            byteArrayOf(NfcByte.CMD_GET_VERSION.toByte())
        else byteArrayOf(NfcByte.N2_GET_VERSION.toByte())
        return transceive(command)
    }

    @Throws(IOException::class)
    fun readPages(pageOffset: Int): ByteArray? {
        return tagMifare?.readPages(pageOffset) ?: if (null != tagNfcA) {
            validatePageIndex(pageOffset)
            //checkConnected();
            val cmd = byteArrayOf(
                NfcByte.CMD_READ.toByte(), pageOffset.toByte()
            )
            tagNfcA.transceive(cmd)
        } else {
            null
        }
    }

    @Throws(IOException::class)
    fun writePage(pageOffset: Int, data: ByteArray) {
        tagMifare?.writePage(pageOffset, data) ?:
        if (null != tagNfcA) {
            validatePageIndex(pageOffset)
            //m_nfcA.checkConnected();
            val cmd = ByteArray(data.size + 2)
            cmd[0] = NfcByte.CMD_WRITE.toByte()
            cmd[1] = pageOffset.toByte()
            System.arraycopy(data, 0, cmd, 2, data.size)
            tagNfcA.transceive(cmd)
        } else {
            throw IOException()
        }
    }

    @Suppress("unused")
    val bankCount: ByteArray?
        get() {
            val req = ByteArray(1)
            req[0] = NfcByte.N2_BANK_COUNT.toByte()
            return transceive(req)
        }

    fun readSignature(isGeneric: Boolean): ByteArray? {
        val command = if (isGeneric)
            byteArrayOf(NfcByte.CMD_READ_SIG.toByte(), 0x00.toByte())
        else byteArrayOf(NfcByte.N2_READ_SIG.toByte())
        return transceive(command)
    }

    fun setBankCount(count: Int) {
        transceive(byteArrayOf(NfcByte.N2_SET_BANKCOUNT.toByte(), (count and 0xFF).toByte()))
    }

    fun activateBank(bank: Int) {
        transceive(byteArrayOf(NfcByte.N2_ACTIVATE_BANK.toByte(), (bank and 0xFF).toByte()))
    }

    fun initFirmware() {
        transceive(
            byteArrayOf(
                0xFFF4.toByte(),
                0x49.toByte(),
                0xFF9B.toByte(),
                0xFF99.toByte(),
                0xFFC3.toByte(),
                0xFFDA.toByte(),
                0x57.toByte(),
                0x71.toByte(),
                0x0A.toByte(),
                0x64.toByte(),
                0x4A.toByte(),
                0xFF9E.toByte(),
                0xFFF8.toByte(),
                NfcByte.CMD_WRITE.toByte(),
                NfcByte.CMD_READ.toByte(),
                0xFFD9.toByte()
            )
        )
    }

    private interface IFastRead {
        fun doFastRead(start: Int, end: Int, bank: Int): ByteArray?
    }

    private interface IFastWrite {
        fun doFastWrite(addr: Int, bank: Int, data: ByteArray): Boolean
    }

    fun fastRead(startAddr: Int, endAddr: Int): ByteArray? {
        return internalFastRead(object : IFastRead {
            override fun doFastRead(start: Int, end: Int, bank: Int): ByteArray? {
                return transceive(
                    byteArrayOf(
                        NfcByte.CMD_FAST_READ.toByte(),
                        (start and 0xFF).toByte(),
                        (end and 0xFF).toByte()
                    )
                )
            }
        }, startAddr, endAddr, 0)
    }

    fun amiiboFastRead(startAddr: Int, endAddr: Int, bank: Int): ByteArray? {
        return internalFastRead(object : IFastRead {
            override fun doFastRead(start: Int, end: Int, bank: Int): ByteArray? {
                return transceive(
                    byteArrayOf(
                        NfcByte.N2_FAST_READ.toByte(),
                        (start and 0xFF).toByte(),
                        (end and 0xFF).toByte(),
                        (bank and 0xFF).toByte()
                    )
                )
            }
        }, startAddr, endAddr, bank)
    }

    private fun internalFastRead(
        method: IFastRead, startAddr: Int, endAddr: Int, bank: Int
    ): ByteArray? {
        if (endAddr < startAddr) return null
        val resp = ByteArray((endAddr - startAddr + 1) * 4)
        val maxReadLength = maxTransceiveLength / 4 - 1
        if (maxReadLength < 1) return null
        val snippetByteSize = maxReadLength * 4
        var startSnippet = startAddr
        var i = 0
        while (startSnippet <= endAddr) {
            var endSnippet = startSnippet + maxReadLength - 1
            if (endSnippet > endAddr) endSnippet = endAddr
            val respSnippet = method.doFastRead(startSnippet, endSnippet, bank) ?: return null
            if (respSnippet.size != (endSnippet - startSnippet + 1) * 4) return null
            if (respSnippet.size == resp.size) return respSnippet
            System.arraycopy(respSnippet, 0, resp, i * snippetByteSize, respSnippet.size)
            startSnippet += maxReadLength
            i++
        }
        return resp
    }

    private fun internalWrite(method: IFastWrite, addr: Int, bank: Int, data: ByteArray): Boolean {
        val query = ByteArray(4)
        for (i in 0 until data.size / 4) {
            System.arraycopy(data, i * 4, query, 0, 4)
            if (!method.doFastWrite(addr + i, bank, query)) {
                return false
            }
        }
        return true
    }

    fun amiiboWrite(addr: Int, bank: Int, data: ByteArray?): Boolean {
        return if (null != data && data.size % 4 == 0) {
            internalWrite(object : IFastWrite {
                override fun doFastWrite(addr: Int, bank: Int, data: ByteArray): Boolean {
                    val req = ByteArray(3)
                    req[0] = NfcByte.N2_WRITE.toByte()
                    req[1] = (addr and 0xFF).toByte()
                    req[2] = (bank and 0xFF).toByte()
                    return try {
                        transceive(req.plus(data.copyOfRange(0, 4)))
                        true
                    } catch (e: Exception) {
                        false
                    }
                }
            }, addr, bank, data)
        } else false
    }

    private fun internalFastWrite(
        method: IFastWrite, startAddr: Int, bank: Int, data: ByteArray
    ): Boolean {
        var snippetByteSize = 16
        val endAddr = startAddr + data.size / 4
        var startSnippet = startAddr
        var i = 0
        while (startSnippet <= endAddr) {
            if (startSnippet + 4 >= endAddr) snippetByteSize = data.size % snippetByteSize
            if (snippetByteSize == 0) return true
            val query = ByteArray(snippetByteSize)
            System.arraycopy(data, i, query, 0, snippetByteSize)
            if (!method.doFastWrite(startSnippet, bank, query)) return false
            startSnippet += 4
            i += snippetByteSize
        }
        return true
    }

    fun amiiboFastWrite(addr: Int, bank: Int, data: ByteArray?): Boolean {
        return if (null == data) {
            false
        } else internalFastWrite(object : IFastWrite {
            override fun doFastWrite(addr: Int, bank: Int, data: ByteArray): Boolean {
                val req = ByteArray(4)
                req[0] = NfcByte.N2_FAST_WRITE.toByte()
                req[1] = (addr and 0xFF).toByte()
                req[2] = (bank and 0xFF).toByte()
                req[3] = (data.size and 0xFF).toByte()
                return try {
                    transceive(req.plus(data))
                    true
                } catch (e: Exception) {
                    false
                }
            }
        }, addr, bank, data)
    }

    fun amiiboLock() {
        transceive(byteArrayOf(NfcByte.N2_LOCK.toByte()))
    }

    fun amiiboPrepareUnlock(): ByteArray? {
        return transceive(byteArrayOf(NfcByte.N2_UNLOCK_1.toByte()))
    }

    fun amiiboUnlock() {
        transceive(byteArrayOf(NfcByte.N2_UNLOCK_2.toByte()))
    }

    companion object {

        private infix fun Short.equals(i: Int): Boolean = this == i.toShort()

        private const val NXP_MANUFACTURER_ID = 0x04
        private const val MAX_PAGE_COUNT = 256

        private fun validatePageIndex(pageIndex: Int) {
            // Do not be too strict on upper bounds checking, since some cards
            // may have more addressable memory than they report.
            // Note that issuing a command to an out-of-bounds block is safe - the
            // tag will wrap the read to an addressable area. This validation is a
            // helper to guard against obvious programming mistakes.
            if (pageIndex < 0 || pageIndex >= MAX_PAGE_COUNT)
                throw IndexOutOfBoundsException("page out of bounds: $pageIndex")
        }

        private fun getMifareUltralight(tag: Tag?): NTAG215? {
            return MifareUltralight.get(tag)?.let {
                NTAG215(it)
            }
        }

        private fun getNfcA(tag: Tag?): NTAG215? {
            return NfcA.get(tag)?.let {
                if (it.sak equals 0x00 && it.tag.id[0].toInt() == NXP_MANUFACTURER_ID)
                    NTAG215(it)
                else
                    null
            }
        }

        @Throws(IOException::class)
        fun getBlind(tag: Tag?): NTAG215 {
            return try {
                NTAG215(NfcA.get(tag)).apply { connect() }
            } catch (ex: IOException) {
                Debug.warn(ex)
                null
            } ?: throw IOException(TagMo.appContext.getString(R.string.error_tag_unavailable))
        }

        @Throws(IOException::class)
        operator fun get(tag: Tag?): NTAG215? {
            return try {
                getMifareUltralight(tag)?.apply { connect() }
            } catch (ex: IOException) {
                Debug.warn(ex)
                null
            } ?: getNfcA(tag)?.apply { connect() }
        }
    }
}